### PR TITLE
Add custom view for other references results

### PR DESCRIPTION
**Pull request by michelleangela**
- contribute a new view to references-view container that displays other references results
- add icons for references types

### CODE REVIEW

File: Extension/src/LanguageServer/extension.ts [hunk 1]
```diff
@@ -1131,6 +1133,13 @@ function onRescanWorkspace(): void {
     clients.forEach(client => client.rescanFolder());
 }
 
+function onShowRefCommand(arg?: ReferenceItem): void {
+    const { location } = arg;
+    vscode.window.showTextDocument(location.uri, {
+        selection: location.range.with({ end: location.range.start })
```
Comment (bobbrow) [inline Extension/src/LanguageServer/extension.ts]: arg is an optional param.  Should it be validated before it is dereferenced?

File: Extension/src/LanguageServer/references.ts [hunk 2]
```diff
@@ -129,11 +131,26 @@ export class ProgressHandler {
         this.client = client;
     }
 
+    initilizeViews(): void {
```
Comment (bobbrow) [inline Extension/src/LanguageServer/references.ts]: [sp] `initialize`

File: Extension/src/LanguageServer/references.ts [hunk 3]
```diff
@@ -93,12 +94,14 @@ function convertReferenceTypeToString(referenceType: ReferenceType, isReferences
...
 }
-
 export class ProgressHandler {
     private client: DefaultClient;
     private disposables: vscode.Disposable[] = [];
 
+    // TODO: move views to class that manages view
     private referencesChannel: vscode.OutputChannel;
+    private findAllRefsView: FindAllRefsView;
+    private viewsInitilized: boolean = false;
```
Comment (bobbrow) [inline Extension/src/LanguageServer/references.ts]: [sp] `viewsInitialized`

File: Extension/src/LanguageServer/referencesModel.ts [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+                fileItem.addReference(reference);
+            }
+        }
+    }
+}
+
+export class ReferenceTypeItem {
+    private files: FileItem[] = [];
+
+    constructor(
```
Comment (bobbrow) [inline Extension/src/LanguageServer/referencesModel.ts]: formatting?

File: Extension/src/LanguageServer/referencesModel.ts [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+            file = new FileItem(uri, fileName);
+            this.files.push(file);
+        }
+        return file;
+    }
+
+    private indexOfFile(fileName: string): number {
+        return this.files.findIndex(function(item): boolean {
+            return item.name === fileName;
+         });
```
Comment (bobbrow) [inline Extension/src/LanguageServer/referencesModel.ts]: nit: indent looks off by one

File: Extension/src/LanguageServer/referencesProvider.ts [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+ * Copyright (c) Microsoft Corporation. All Rights Reserved.
+ * See 'LICENSE' in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+'use strict';
+import * as vscode from 'vscode';
+import * as util from '../common';
+import { Model, FileItem, ReferenceItem, ReferenceTypeItem } from './referencesModel';
+import { ReferenceInfo, ReferenceType, convertReferenceTypeToString } from './references';
+
+function getReferneceTypeIconPath(referenceType: ReferenceType): vscode.ThemeIcon {
```
Comment (bobbrow) [inline Extension/src/LanguageServer/referencesProvider.ts]: [sp] `getReferenceTypeIconPath`

**Review by Colengms — APPROVED** [review]

File: Extension/src/LanguageServer/referencesProvider.ts [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+        if (element instanceof ReferenceTypeItem) {
+            return element.getFiles();
+        }
+
+        return this.references.items;
+    }
+}
+
+// TODO: add provider for rename
```
Comment (bobbrow) [inline Extension/src/LanguageServer/referencesProvider.ts:95]: Would this be the right file to put 'rename' stuff in?  This is referencesProvider.ts.  Perhaps renaming the file or putting the provider in a different file would be better.

File: Extension/src/LanguageServer/referencesView.ts [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    setData(results: ReferenceInfo[]): void {
+        this.referenceViewProvider.setModel(results);
+    }
+
+    clearData(): void {
+        this.referenceViewProvider.clear();
+    }
+}
+
+// TODO: create class to manage displaying data to rename view
```
Comment (bobbrow) [inline Extension/src/LanguageServer/referencesView.ts:36]: Similar comment as above.  'rename' stuff should probably get its own file.

**Review by bobbrow — COMMENTED** [review]
General feedback.

File: Extension/src/LanguageServer/extension.ts (hunk 1 above)
Comment (michelleangela) [reply Extension/src/LanguageServer/extension.ts]: There should be check on "arg"

File: Extension/src/LanguageServer/referencesProvider.ts (hunk 7 above)
Comment (michelleangela) [reply Extension/src/LanguageServer/referencesProvider.ts:95]: Pending design of rename and separating of current code that's used for reference results and rename. Planning to check this in first for Colen and I to collaborate on.

File: Extension/src/LanguageServer/referencesView.ts (hunk 8 above)
Comment (michelleangela) [reply Extension/src/LanguageServer/referencesView.ts:36]: Pending design for rename and further separate of code for references results and rename.

File: Extension/package.json [hunk 9]
```diff
@@ -43,6 +43,15 @@
   ],
   "main": "./dist/main",
   "contributes": {
+    "views": {
+      "references-view": [
+        {
+          "id": "CppReferencesView",
+          "name": "%c_cpp.contributes.views.cppReferencesView.title%",
+          "when": "cppReferenceTypes:hasResults && reference-list.hasResult"
```
Comment (sean-mcmanus) [inline Extension/package.json:51]: I think this should be ||, right?

File: Extension/src/LanguageServer/extension.ts [hunk 10]
```diff
@@ -1131,6 +1133,15 @@ function onRescanWorkspace(): void {
     clients.forEach(client => client.rescanFolder());
 }
 
+function onShowRefCommand(arg?: ReferenceItem): void {
+    if (arg) {
+        const { location } = arg;
+        vscode.window.showTextDocument(location.uri, {
+            selection: location.range.with({ end: location.range.start })
```
Comment (sean-mcmanus) [inline Extension/src/LanguageServer/extension.ts:1140]: We could send back the end or the source string to get the full range later.

File: Extension/src/LanguageServer/references.ts [hunk 11]
```diff
@@ -218,58 +235,65 @@ export class ProgressHandler {
...
+
         switch (notificationBody.referencesProgress) {
             case ReferencesProgress.StartedRename:
             case ReferencesProgress.Started:
-                this.referencesStartedWhileTagParsing = this.client.IsTagParsing;
-                let mode: ReferencesCommandMode = notificationBody.referencesProgress === ReferencesProgress.StartedRename ? ReferencesCommandMode.Rename :
-                    (this.visibleRangesDecreased && (Date.now() - this.visibleRangesDecreasedTicks < this.ticksForDetectingPeek) ?
-                    ReferencesCommandMode.Peek : ReferencesCommandMode.Find);
-                this.client.setReferencesCommandMode(mode);
                 if (this.client.ReferencesCommandMode === ReferencesCommandMode.Peek) {
```
Comment (sean-mcmanus) [inline Extension/src/LanguageServer/references.ts:293]: Why is this not in the handleProgressStarted method?

File: Extension/src/LanguageServer/referencesProvider.ts [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+ * ------------------------------------------------------------------------------------------ */
+'use strict';
+import * as vscode from 'vscode';
+import * as util from '../common';
+import { Model, FileItem, ReferenceItem, ReferenceTypeItem } from './referencesModel';
+import { ReferenceInfo, ReferenceType, convertReferenceTypeToString } from './references';
+
+function getReferenceTypeIconPath(referenceType: ReferenceType): vscode.ThemeIcon {
+    // TODO: return icon path for light and dark themes based on reference type
+    switch (referenceType) {
```
Comment (sean-mcmanus) [inline Extension/src/LanguageServer/referencesProvider.ts:13]: ConfirmationInProgress is missing. The icon should change depending on if the operation is canceled or not. VS doesn't appear to use any icon for this though, even though it's a type.

**Comment by sean-mcmanus** [comment]
Do you know where we are going to expose the "Warning message" about the tag parsing not being finished? I see that it's still shown in the C/C++ references output window -- can the warning appear in the tree view if that is used? If not, it seems like it should be moved to the regular C/C++ output window if the references output isn't being used.

**Review by sean-mcmanus — APPROVED** [review]

File: Extension/package.json (hunk 9 above)
Comment (sean-mcmanus) [reply Extension/package.json:51]: Or just cppReferenceTypes:hasResults ? If there isn't any reason to show the cppReferenceTypes if there are only confirmed references.

**Review by sean-mcmanus — CHANGES_REQUESTED** [review]
"when": "cppReferenceTypes:hasResults ?

**Comment by michelleangela** [comment]
The custom view is only shown when there are confirmed results and there are other reference types. Sometimes, there are only confirmed results, so there is nothing to output in the custom references view.

File: Extension/src/LanguageServer/referencesModel.ts [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+            if (index < 0) {
+                const uri: vscode.Uri = vscode.Uri.file(r.file);
+                fileItem = new FileItem(uri, r.file);
+                this.items.push(fileItem);
+            } else {
+                fileItem = this.items[index];
+            }
+
+            // Add reference to file
+            let isFileReference: boolean = (r.position.line === 0 && r.position.character === 0);
```
Comment (sean-mcmanus) [inline Extension/src/LanguageServer/referencesModel.ts:29]: Can you handle "file references"? Those are cases where the line/column are not known yet, in which case, it seems like it should add a link to just the start of the file.

Oh, it looks like it adds the file, but just leaves out the reference...that seems fine.

**Comment by michelleangela** [comment]
> 's still shown in the C/C++ references output window -- can the warning appear in the tree view if that is used? If not, it seems like it should be moved to the regular C/C

I don't know where to place this warning yet.

File: Extension/src/LanguageServer/referencesProvider.ts (hunk 12 above)
Comment (michelleangela) [reply Extension/src/LanguageServer/referencesProvider.ts:13]: Will add icon for ConfirmationInProgress later.

**Review by sean-mcmanus — APPROVED** [review]

File: Extension/package.json (hunk 9 above)
Comment (michelleangela) [reply Extension/package.json:51]: Verified that reference-list.hasResult is also true if there are no confirmed references or if you cancel "find all references" while in progress.

File: Extension/src/LanguageServer/extension.ts (hunk 10 above)
Comment (michelleangela) [reply Extension/src/LanguageServer/extension.ts:1140]: Having the full range of the symbol result would mean we will highlight in the editor. A selection of a confirmed reference already highlights the symbol. Should we highlight two different selections?

File: Extension/src/LanguageServer/references.ts (hunk 11 above)
Comment (michelleangela) [reply Extension/src/LanguageServer/references.ts:293]: I wanted to separate telemetry stuff from the logic that processes the data.

File: Extension/src/LanguageServer/referencesModel.ts (hunk 13 above)
Comment (michelleangela) [reply Extension/src/LanguageServer/referencesModel.ts:29]: We still need to handle file references.

**Pull request merged**